pystiche/papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,200 @@
+    "ulyanov_et_al_2016_perceptual_loss",
+    "ulyanov_et_al_2016_dataset",
+    "ulyanov_et_al_2016_image_loader",
+    "ulyanov_et_al_2016_training",
+    "ulyanov_et_al_2016_images",
+    "ulyanov_et_al_2016_style_generation",
+    "ulyanov_et_al_2016_texture_generation",
+]
+
+
+def ulyanov_et_al_2016_training(
+    content_image_loader: DataLoader,
+    style: Union[str, torch.Tensor],
+    impl_params: bool = True,
+    instance_norm: bool = True,
+    stylization: bool = True,
+    transformer: Optional[UlyanovEtAl2016Transformer] = None,
+    criterion: Optional[UlyanovEtAl2016PerceptualLoss] = None,
+    lr_scheduler: Optional[ExponentialLR] = None,
+    num_epochs: Optional[int] = None,
+    get_optimizer: Optional[Callable[[UlyanovEtAl2016Transformer], Optimizer]] = None,
+    quiet: bool = False,
+    logger: Optional[OptimLogger] = None,
+    log_fn: Optional[
+        Callable[[int, Union[torch.Tensor, pystiche.LossDict], float, float], None]
+    ] = None,
+):
+    if isinstance(style, str):
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        images = ulyanov_et_al_2016_images(download=False)
+        style_image = images[style].read(device=device)
+    else:
+        style_image = style
+        device = style_image.device
+
+    if transformer is None:
+        transformer = ulyanov_et_al_2016_transformer(
+            impl_params=impl_params,
+            instance_norm=instance_norm,
+            stylization=stylization,
+        )
+        transformer = transformer.train()
+    transformer = transformer.to(device)
+
+    if criterion is None:
+        criterion = ulyanov_et_al_2016_perceptual_loss(
+            impl_params=impl_params,
+            instance_norm=instance_norm,
+            stylization=stylization,
+        )
+        criterion = criterion.eval()
+    criterion = criterion.to(device)
+
+    if lr_scheduler is None:
+        if get_optimizer is None:
+            get_optimizer = ulyanov_et_al_2016_optimizer
+        optimizer = get_optimizer(transformer)
+
+        lr_scheduler = ulyanov_et_al_2016_lr_scheduler(
+            optimizer, impl_params=impl_params,
+        )
+
+    if num_epochs is None:
+        if impl_params:
+            if instance_norm:
+                num_epochs = 25
+            else:
+                num_epochs = 10 if stylization else 5
+        else:
+            num_epochs = 10
+
+    style_transform = ulyanov_et_al_2016_style_transform(
+        impl_params=impl_params, instance_norm=instance_norm
+    )
+    style_transform = style_transform.to(device)
+    preprocessor = ulyanov_et_al_2016_preprocessor()
+    preprocessor = preprocessor.to(device)
+    style_image = style_transform(style_image)
+    style_image = preprocessor(style_image)
+    style_image = batch_up_image(style_image, loader=content_image_loader)
+    criterion.set_style_image(style_image)
+
+    def criterion_update_fn(input_image, criterion):
```

Comment:
If `stylization` is `false`, nothing is done in the function. Is `pass` the right command? 

```python
    if stylization:
        def criterion_update_fn(input_image, criterion):
            criterion.set_content_image(preprocessor(input_image))
    else:
        def criterion_update_fn(input_image, criterion):
            pass
```

Reply:
Since in python the indentation is part of the language, you need to provide at least a single indented command after commands like `def`. If nothing should be done inside the block, the `pass` keyword is the correct choice.